inst/shiny/sandbox/server.R

```diff
@@ -0,0 +1,199 @@
+  
+  observe({
+    shinyjs::disable("remove_module")
+  })
+  
+  ## UI only works with debug = TRUE for some reason (!)
+  debug <- TRUE
+  bs4dash <- getOption("facile.bs4dash")
+  options(facile.bs4dash = FALSE)
+  on.exit(options(facile.bs4dash = bs4dash))
+  
+  x <- eventReactive(input$dataset, {
+    d <- switch(req(input$dataset),
+           "TCGA" = FacileData:::exampleFacileDataSet(),
+           results_stack()[results_stack()$id == input$dataset, "result", drop = TRUE][[1]]
+    )
+    if (is(d, "ReactiveFacileAnalysisResultContainer")) {
+      .x <- FacileAnalysis::faro(d)
+    } else {
+      .x <- d
+    }
+    .x
+  })
+  
+  analysisModule <- reactive({
+    switch(req(input$analysis),
+           "filter" = FacileShine::filteredReactiveFacileDataStore,
+           "fdge" = FacileAnalysis::fdgeAnalysis,
+           "fpca" = FacileAnalysis::fpcaAnalysis,
+           "ffsea" = FacileAnalysis::ffseaAnalysis,
+           NULL
+    )
+  })
+
+  analysisUI <- reactive({
+    switch(req(input$analysis),
+           "filter" = FacileShine::filteredReactiveFacileDataStoreUI,
+           "fdge" = FacileAnalysis::fdgeAnalysisUI,
+           "fpca" = FacileAnalysis::fpcaAnalysisUI,
+           "ffsea" = FacileAnalysis::ffseaAnalysisUI,
+           NULL
+    )
+  })
+    
+  observeEvent(input$add_module, {
+    req(input$analysis != "none")
+    
+    # store the id of the newly added module using the 
+    # value of the actionButton to make it unique
+    module_id <- paste0("id_", input$add_module)
+    if (debug) print(paste0("this module is ", module_id))
+    module_stack(c(module_id, module_stack()))
+    
+    ui.content <- analysisUI()(ifelse(isolate(req(input$analysis)) == "filter","ds","analysis"), debug = debug)
+    
+    ui <- tagList(
+      ui.content
+    )
+    
+    ui_with_id <- module_UI(module_id, ui)
+    
+    ## NOTE: immediate = TRUE is necessary!
+    insertUI("#gadget_container", "afterEnd", ui_with_id, immediate = TRUE)
+    
+    shinyjs::disable("add_module")
+    shinyjs::enable("remove_module")
+    
+    isolate(module_res())
+  })
+  
+  observeEvent(input$remove_module, {
+    if (NROW(req(module_stack())) > 0) {
+      if (debug) {
+        print(paste0("removing module ", module_stack()[1]))
+      }
+      removeUI(paste0("#", module_stack()[1]))
+    }
+    remove_shiny_inputs(module_stack()[1], input)
+    remove_observers(module_stack()[1], session)
+    module_stack(module_stack()[-1])
+    
+    shinyjs::enable("add_module")
+    shinyjs::disable("remove_module")
+  })
+  
+  ## this logic should be isolated into a function
+  rfds <- reactive({
+    req(input$dataset)
+    req(input$analysis != "none")
+    
+    .x <- x()
+    if (is(.x, "facile_frame")) {
+      if (debug) print("facile_frame")
+      fds. <- FacileData::fds(.x)
+      samples. <- .x
+      sample.filter <- FALSE
+      restrict_samples <- samples.
+    } else if (is(.x, "ReactiveFacileDataStore")) {
```

Comment:
This logic is growing a bit too big, but I think this is the correct way to extract out the relevant parts - feel free to correct me, I'm still learning how these parts work.